File: frontend/src/api/nlp-search.ts
// Added: NLP search API client for TMAIL-135
// PURPOSE: Provides functions for AI-powered natural language email search and history management

import { apiClient } from './client';

/// PURPOSE: Structured search parameters parsed from a natural language query by the AI
export interface ParsedSearchParams {
  from?: string;
  to?: string;
  subject?: string;
  keywords: string[];
  date_from?: string;
  date_to?: string;
  folder?: string;
  has_attachment?: boolean;
}

/// PURPOSE: A single message result from NLP search
export interface NlpSearchResultItem {
  folder: string;
  uid: number;
  subject: string | null;
  from: string | null;
  date: string | null;
}

/// PURPOSE: Full NLP search response including parsed params and results
export interface NlpSearchResult {
  query: string;
  parsed_params: ParsedSearchParams;
  result_count: number;
  results: NlpSearchResultItem[];
}

/// PURPOSE: A single entry from the NLP search history
export interface NlpSearchHistoryEntry {
  id: string;
  user_id: string;
  query_text: string;
  parsed_params: ParsedSearchParams;
  result_count: number;
  created_at: string;
}

// PURPOSE: Execute a natural language search query via AI parsing
// CONSTRAINTS: Requires an active AI config on the backend
export async function nlpSearch(query: string): Promise<NlpSearchResult> {
  return apiClient.post<NlpSearchResult>('/search/nlp', { query });
}

// PURPOSE: List the user's NLP search history (most recent first, max 50)
export async function listNlpHistory(): Promise<NlpSearchHistoryEntry[]> {
  return apiClient.get<NlpSearchHistoryEntry[]>('/search/nlp/history');
}

// PURPOSE: Clear all NLP search history for the current user
export async function clearNlpHistory(): Promise<{ deleted: number; message: string }> {
  return apiClient.delete<{ deleted: number; message: string }>('/search/nlp/history');
}

File: mobile/lib/models/nlp_search.dart
// Added: NLP search data models for TMAIL-147
// PURPOSE: Dart mirrors of backend ParsedSearchParams / NlpSearchResponse /
//          NlpSearchResultItem / NlpSearchHistory so the search screen can
//          surface what the AI parsed and the user's prior queries.
// EXTERNAL: Maps to backend/src/models/nlp_search.rs

/// PURPOSE: AI-parsed structured parameters extracted from a free-text query.
/// NOTE: Every field is optional — the AI may extract a subset.
class ParsedSearchParams {
  final String? from;
  final String? to;
  final String? subject;
  final List<String> keywords;
  final String? dateFrom;
  final String? dateTo;
  final String? folder;
  final bool? hasAttachment;

  const ParsedSearchParams({
    this.from,
    this.to,
    this.subject,
    this.keywords = const [],
    this.dateFrom,
    this.dateTo,
    this.folder,
    this.hasAttachment,
  });

  factory ParsedSearchParams.fromJson(Map<String, dynamic> json) {
    return ParsedSearchParams(
      from: json['from'] as String?,
      to: json['to'] as String?,
      subject: json['subject'] as String?,
      keywords: (json['keywords'] as List<dynamic>?)
              ?.map((k) => k.toString())
              .toList() ??
          const [],
      dateFrom: json['date_from'] as String?,
      dateTo: json['date_to'] as String?,
      folder: json['folder'] as String?,
      hasAttachment: json['has_attachment'] as bool?,
    );
  }

  // PURPOSE: Used by the parsed-params banner to decide whether to render at all.
  bool get isEmpty =>
      from == null &&
      to == null &&
      subject == null &&
      keywords.isEmpty &&
      dateFrom == null &&
      dateTo == null &&
      folder == null &&
      hasAttachment == null;
}

/// PURPOSE: A single NLP search result row. Distinct from MobileMessageSummary —
///          backend's NlpSearchResultItem does NOT carry is_read/is_flagged/
///          has_attachment, so we keep a minimal struct that only models what
///          the endpoint actually returns.
class NlpSearchResultItem {
  final String folder;
  final int uid;
  final String? subject;
  final String? from;
  final String? date;

  const NlpSearchResultItem({
    required this.folder,
    required this.uid,
    this.subject,
    this.from,
    this.date,
  });

  factory NlpSearchResultItem.fromJson(Map<String, dynamic> json) {
    return NlpSearchResultItem(
      folder: json['folder'] as String,
      uid: json['uid'] as int,
      subject: json['subject'] as String?,
      from: json['from'] as String?,
      date: json['date'] as String?,
    );
  }
}

/// PURPOSE: Wraps the full /api/search/nlp response so callers see both the
///          parsed params (for the banner) and the result rows.
class NlpSearchResponse {
  final String query;
  final ParsedSearchParams parsedParams;
  final int resultCount;
  final List<NlpSearchResultItem> results;

  const NlpSearchResponse({
    required this.query,
    required this.parsedParams,
    required this.resultCount,
    required this.results,
  });

  factory NlpSearchResponse.fromJson(Map<String, dynamic> json) {
    return NlpSearchResponse(
      query: json['query'] as String? ?? '',
      parsedParams: ParsedSearchParams.fromJson(
        (json['parsed_params'] as Map<String, dynamic>?) ?? const {},
      ),
      resultCount: json['result_count'] as int? ?? 0,
      results: (json['results'] as List<dynamic>?)
              ?.map((r) => NlpSearchResultItem.fromJson(r as Map<String, dynamic>))
              .toList() ??
          const [],
    );
  }
}

/// PURPOSE: One entry from GET /api/search/nlp/history. The backend stores
///          parsed_params as JSONB; we round-trip it through ParsedSearchParams
///          so the history sheet can show the same banner if needed.
class NlpSearchHistoryEntry {
  final String id;
  final String queryText;
  final ParsedSearchParams parsedParams;
  final int resultCount;
  final DateTime createdAt;

  const NlpSearchHistoryEntry({
    required this.id,
    required this.queryText,
    required this.parsedParams,
    required this.resultCount,
    required this.createdAt,
  });

  factory NlpSearchHistoryEntry.fromJson(Map<String, dynamic> json) {
    final parsedRaw = json['parsed_params'];
    final parsedMap = parsedRaw is Map<String, dynamic>
        ? parsedRaw
        : const <String, dynamic>{};
    return NlpSearchHistoryEntry(
      id: json['id'].toString(),
      queryText: json['query_text'] as String? ?? '',
      parsedParams: ParsedSearchParams.fromJson(parsedMap),
      resultCount: (json['result_count'] as num?)?.toInt() ?? 0,
      createdAt: DateTime.tryParse(json['created_at']?.toString() ?? '') ??
          DateTime.fromMillisecondsSinceEpoch(0),
    );
  }
}
